src/validation.cpp

```diff
@@ -3681,6 +3702,51 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
     return true;
 }
 
+bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
+{
+    LOCK(cs_main);
+
+    CCoinsViewCache cache(view);
+
+    uint256 hashBest = view->GetBestBlock();
+    uint256 hashUpto = view->GetUptoBlock();
+
+    if (hashUpto.IsNull() || hashBest == hashUpto) return true;
+
+    if (mapBlockIndex.count(hashUpto) == 0 ||
+        (!hashBest.IsNull() && mapBlockIndex.count(hashBest) == 0)) {
+        return error("ReplayBlocks(): chainstate boundaries not in block index");
+    }
+    auto pindexUpto = mapBlockIndex[hashUpto];
+
+    int nHeight = 1; // Skip the genesis block
+    if (mapBlockIndex.count(hashBest)) {
+        auto pindexBest = mapBlockIndex[hashBest];
+        if (pindexUpto->GetAncestor(pindexBest->nHeight) != pindexBest) {
+            return error("ReplayBlocks(): chainstate tip does not derive from final boundary");
```

Comment:
That's a great point. I hadn't considered that in the case of a reorg the set of partially written changes may include things from multiple branches. It seems solvable by allowing the 'upto' blocks to be a list of tip hashes, and then at reply time choose which ones to undo and which ones to replay. I think that's a problem for later, but it makes sense to have a comment about it.